Dockerfile

```diff
@@ -0,0 +1,47 @@
+# CUDA 12.2
+FROM nvcr.io/nvidia/pytorch:23.10-py3
+
+# if you get errors building TE or Apex, decrease this to 4
+ARG MAX_JOBS=8
+# needed in case git complains that it can't detect a valid email, this email is fake but works
+RUN git config --global user.email "worker@nvidia.com"
+
+WORKDIR /opt
+
+# install TransformerEngine
+RUN pip uninstall -y transformer-engine
+RUN pip install --upgrade git+https://github.com/NVIDIA/TransformerEngine.git@release_v1.1
+
+# install latest apex
+RUN pip uninstall -y apex
+RUN git clone https://github.com/NVIDIA/apex.git
+RUN cd apex && pip install -v --disable-pip-version-check --no-cache-dir --no-build-isolation --config-settings "--build-option=--cpp_ext" --config-settings "--build-option=--cuda_ext" --config-settings "--build-option=--distributed_adam" --config-settings "--build-option=--permutation_search" --config-settings "--build-option=--bnp" --config-settings "--build-option=--xentropy" --config-settings "--build-option=--focal_loss" --config-settings "--build-option=--index_mul_2d" --config-settings "--build-option=--deprecated_fused_adam" --config-settings "--build-option=--fast_layer_norm" --config-settings "--build-option=--fmha" --config-settings "--build-option=--fast_multihead_attn" --config-settings "--build-option=--transducer" --config-settings "--build-option=--cudnn_gbn" --config-settings "--build-option=--fused_conv_bias_relu" .
+
+# place any util pkgs here
+RUN pip install --upgrade-strategy only-if-needed nvidia-pytriton==0.4.1
+RUN pip install -U --no-deps protobuf==4.24.4
+RUN pip install --upgrade-strategy only-if-needed jsonlines
+
+# NeMo
+RUN git clone -b r1.22.0 https://github.com/NVIDIA/NeMo.git
+RUN cd NeMo && git cherry-pick -X theirs fa8d416793d850f4ce56bea65e1fe28cc0d092c0
+WORKDIR /opt/NeMo
+# fixes a bug in NeMo when used with latest apex
+RUN sed -i 's/_fast_layer_norm(x, self.weight + 1, self.bias, self.epsilon)/_fast_layer_norm(x, self.weight + 1, self.bias, self.epsilon, False)/g' nemo/collections/nlp/modules/common/megatron/layer_norm_1p.py
+RUN sed -i 's/shutil.rmtree(ckpt_to_dir(filepath))/shutil.rmtree(ckpt_to_dir(filepath), ignore_errors=True)/g' nemo/collections/nlp/parts/nlp_overrides.py
+RUN sed -i 's/\[all\]/\[nlp\]/g' reinstall.sh
+RUN rm -rf .git && ./reinstall.sh
```

Comment:
is it possible to upstream this onto nemo? if they ever change this file the sed may start to not work

Reply:
Sadly not for the `_fast_layer_norm` part because Nemo is still using an old version of Apex whereas our Dockerfile uses the latest Apex, hence we need to make that change or else there's an error. Whenever Nemo chooses to upgrade to the latest Apex, they will make that change.

The `all` -> `nlp` we need to change locally as we are reinstalling just the NLP parts, we don't need the ASR and TTS packages.

The shutil.rmtree fix they should add, I'll add a PR for it.

Reply:
Also, these sed commands are quite specifically focused, no wildcards or anything, so if the upstream codes changes at all, these won't find what they're specifically looking for and shouldn't impact anything. I made them to be quite benign in that way.

Reply:
ah i see, does this change only affect the non mcore path? if that is the case I am okay with even just removing these lines. Since aligner technically doesn't support non mcore(though in theory it runs)

Reply:
The `_fast_layer_norm` is only for non-mcore, but the others are necessary for mcore too.

Reply:
ah i see!

so how about this:

line 30: we can remove, because we don't support non-mcore models(and this enforces it)
line 31: this needs to be changed on the nemo side, i've seen customers complain about this
line 32: let's keep it

how does that sound?

Reply:
I have removed line 30. Line 31 we need to keep until https://github.com/NVIDIA/NeMo/pull/8130/files gets merged in and I can cherry pick it in. Line 32 left in as requested.